File: lab1/CoolFilms/CoolFilms/ViewModel/FilmsListViewModel.swift
import SwiftUI
import FirebaseDatabase

class FilmsListViewModel: ObservableObject {
    @Published var films: [Film] = []
    
    private let database = Database.database()
    
    func loadFilms() {
        let filmsRef = database.reference().child("films")
        filmsRef.observe(.value) { snapshot in
            var films: [Film] = []
            
            for child in snapshot.children {
                if let snapshot = child as? DataSnapshot,
                   let filmId = snapshot.key as String?,
                   let filmValue = snapshot.value as? [String: Any] {
                    var film = Film(name: filmValue["name"] as? String ?? "",
                                    categories: filmValue["categories"] as? [String] ?? [],
                                    description: filmValue["description"] as? String ?? "")
                    film.id = filmId
                    films.append(film)
                }
            }
            
            self.films = films
        }
    }
}

File: lab3/lib/viewmodel/films_list_viewmodel.dart
import 'package:flutter/material.dart';
import 'package:firebase_database/firebase_database.dart';

import '../model/film.dart';

class FilmsListViewModel extends ChangeNotifier {
  List<Film> films = [];

  final database = FirebaseDatabase.instance;

  void loadFilms() {
    final filmsRef = database.ref().child('films');
    filmsRef.onValue.listen((event) {
      final snapshot = event.snapshot;
      final films = <Film>[];
      for (final child in snapshot.children) {
        final filmId = child.key;
        final filmValue = child.value as Map<dynamic, dynamic>;
        final film = Film(
            name: filmValue['name'] as String? ?? '',
            categories:
                List<String>.from(filmValue['categories'] as List? ?? []),
            description: filmValue['description'] as String? ?? '',
            id: filmId);
        films.add(film);
      }
      this.films = films;
      notifyListeners();
    });
  }
}
